### PR TITLE
[SPARK-34193][CORE] TorrentBroadcast block manager decommissioning race fix

### PR DESCRIPTION
**Pull request by holdenk**
### What changes were proposed in this pull request?

Allow broadcast blocks to be put during decommissioning since migrations don't apply to them and they may be stored as part of job exec.

### Why are the changes needed?

Potential race condition.

### Does this PR introduce _any_ user-facing change?

Removal of race condition.

### How was this patch tested?

New unit test.

### CODE REVIEW

**Comment by holdenk** [comment]
Hive failures are probably unrelated, jenkins retest this please.

**Comment by holdenk** [comment]
Jenkins retest this please
(kafka failure unrelated).

**Comment by HyukjinKwon** [comment]
cc @Ngone51 who also works on decommissioning and torrent broadcast stuff.

**Comment by Ngone51** [comment]
``` scala
[info]   org.apache.spark.SparkException: Job aborted due to stage failure:
Task 1 in stage 3.0 failed 4 times, most recent failure: Lost task 1.3 in stage 3.0 (TID 8, 192.168.1.57, executor 1):
java.io.IOException: org.apache.spark.storage.BlockSavedOnDecommissionedBlockManagerException:
Block broadcast_2_piece0 cannot be saved on decommissioned executor
[info]   org.apache.spark.SparkException: Job aborted due to stage failure:
Task 1 in stage 3.0 failed 4 times, most recent failure: Lost task 1.3 in stage 3.0 (TID 8, 192.168.1.57, executor 1): java.io.IOException: org.apache.spark.storage.BlockSavedOnDecommissionedBlockManagerException: Block broadcast_2_piece0 cannot be saved on decommissioned executor[info] at
...
```

Usually, the task fails with the max failure times would due to the same reason. But in the case of `BlockSavedOnDecommissionedBlockManagerException`, I'd not expect the 4 times failures are all caused by `BlockSavedOnDecommissionedBlockManagerException` since we don't schedule tasks on the decommissioning executor. (Unless all the task failures always happen before the decommission notice arrives at the driver. But this seems very impossible.)

So to ensure that we still ensure that we don't schedule tasks on the decommissioning executor (or there's really a race condition or it's just a coincidence),  I'd like to know the previous task failures for this task (TID = 8). e.g., what are failures reasons for previous attempts? And which executors are scheduled?

BTW, the change itself lgtm.

**Comment by holdenk** [comment]
> So to double check that we still ensure that we don't schedule tasks on the decommissioning executor (or there's really a race condition or it's just a coincidence), I'd like to know the previous task failures for this task (TID = 8). e.g., what are failures reasons for previous attempts? And which executors are scheduled?

Yes so to be clear I got this while executing in a tight loop of test code wherein I intentionally scheduled the job. I'm not sure this race condition can occur in the "real world" to the point the task actually fails, but we might abort some started computation which is undeseriable (but not a major issue).

**Comment by holdenk** [comment]
cc @attilapiros

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 1]
```diff
@@ -258,6 +258,17 @@ private[spark] class BlockManager(
   @inline final private def isDecommissioning() = {
     decommissioner.isDefined
   }
+
+  @inline final private def checkShouldStore(blockId: BlockId) = {
+    if (isDecommissioning()) {
```
Comment (attilapiros) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala:263]: Nit: I would avoid the nested if.  
```suggestion
    if (isDecommissioning() && !blockId.isBroadcast) {
```

**Review by attilapiros — APPROVED** [review]
just one Nit otherwise LGTM

**Comment by Ngone51** [comment]
Thanks for the explanation @holdenk . +1 to address @attilapiros 's suggestion, otherwise LGTM.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala (hunk 1 above)
Comment (holdenk) [reply core/src/main/scala/org/apache/spark/storage/BlockManager.scala:263]: Oh yeah I did the nested if for branch prediction but yeah I'll collapse it down.

**Comment by HyukjinKwon** [comment]
cc @jiangxb1987 would you mind taking a look and merging this one when you find some time please?

**Comment by holdenk** [comment]
@HyukjinKwon do you want me to wait for @jiangxb1987's review? Personally I'm comfortable with @attilapiros's sign-off.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
I think it's fine to merge and leave it for a post review given that the RC will be cut out soon. I reviewed too and signed off.

Merged to master and branch-3.1.

**Comment by mridulm** [comment]
Late LGTM, thanks for the fix @holdenk !

**Pull request closed without merging**